kernel/src/main/java/org/kframework/parser/binary/BinaryParser.java

```diff
@@ -53,72 +57,100 @@
  */
 public class BinaryParser {
 
-    private static final byte[] MAGIC = {0x7f, 'K', 'A', 'S', 'T'};
+    public static final byte[] MAGIC = {0x7f, 'K', 'A', 'S', 'T'};
 
     public static final int BEGIN = 0, KTOKEN = 1, KAPPLY = 2, KSEQUENCE = 3, KVARIABLE = 4, KREWRITE = 5,
-            INJECTEDKLABEL = 6, END = 7;
+            INJECTEDKLABEL = 6, END = 7, BACK_REFERENCE = 8;
 
     private final ByteBuffer data;
     private final List<String> interns = new ArrayList<>();
+    private final List<K> kInterns = new ArrayList<>();
+
+    private static K[] EMPTY_KLIST = new K[0];
 
     private BinaryParser(ByteBuffer data) {
         this.data = data;
     }
 
-    private K read400() throws IOException {
+    private K read400(boolean _401) throws IOException {
 
         Deque<K> stack = new ArrayDeque<>();
         int type = 0;
         while(type != END) {
             type = data.get();
-            scala.collection.immutable.List<K> items;
+            K[] items;
             int arity;
             switch (type) {
             case KTOKEN:
-                stack.push(KToken(readString(), Sort(readString())));
+                String s = readString();
+                String sort = readString();
+                Map<String, KToken> sortCache = ktokenCache.computeIfAbsent(sort, sort2 -> new HashMap<>());
```

Comment:
basically the rest of this stuff including the code in org.kframework.kore.mini is designed to decrease the memory footprint of terms parsed from binary in java.